### PR TITLE
Collection expressions: empty Span optimization

### PR DESCRIPTION
**Pull request by RikkiGibson**
Part of #68785

### CODE REVIEW

**Review by cston — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Love it

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 1]
```diff
@@ -15790,12 +15795,14 @@ static void Report<T>(MyCollection<T> c)
...
                 verifier.VerifyIL("Program.Empty<T>", """
                     {
-                      // Code size       16 (0x10)
+                      // Code size       15 (0xf)
                       .maxstack  1
-                      IL_0000:  call       "T[] System.Array.Empty<T>()"
-                      IL_0005:  newobj     "System.ReadOnlySpan<T>..ctor(T[])"
-                      IL_000a:  call       "MyCollection<T> MyCollectionBuilder.Create<T>(scoped System.ReadOnlySpan<T>)"
-                      IL_000f:  ret
+                      .locals init (System.ReadOnlySpan<T> V_0)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:15800]: Do we understand why the new codegen involves an additional local and IL instruction?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:15800]: initobj expects an address on the stack. In some cases a new local is introduced in order to provide that.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by CyrusNajmabadi** [comment]
Note: we should do this for 17.8.  This is a very trivial change, but important to our messaging about collections and perf.  Having a core span scenario work well is high on that list.

**Pull request merged**